File: Data Structures/Tree-based-DSAs/Binary-tree/avl-tree.py
""" avl tree implementation in python: """
import sys
#tree node: 

class Node(object):
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None
        self.height = 1

class AVLTree(object):
    def insert_node(self, root, key):
        #find the correct location and insert the node:

        if not root:
            return Node(key)
        elif key < root.key:
            root.left = self.insert_node(root.left, key)
        else:
            root.right = self.insert_node(root.right, key)
        
        #get the height:
        root.height = 1 + max(self.getHeight(root.left), self.getHeight(root.right))

        #update the balance factor and balance the tree:
        balanceFactor = self.get_balance(root)
        if balanceFactor > 1:
            if key < root.left.key:
                return self.rightRotate(root)
            else:
                root.left = self.leftRotate(root.left)
                return self.rightRotate(root)

        if balanceFactor < -1:
            if key > root.right.key:
                return self.leftRotate(root)
            else:
                root.right = self.rightRotate(root.right)
                return self.leftRotate(root)
        #tree is well balanced:
        return root 

    def delete_node(self, root, key):
        #find the node to be deleted and remove it:

        if not root:
            return root
        elif key < root.key:
            root.left = self.delete_node(root.left, key)
        elif key > root.key:
            root.right = self.delete_node(root.right, key)
        else:
            if root.left is None:
                temp = root.right
                root = None
                return temp
            elif root.right is None:
                temp = root.left
                root = None
                return temp
            temp = self.getMinValNode(root.right)
            root.key = temp.key
            root.right = self.delete_node(root.right, temp.key)
        
        if root is None:
            return root

        #get the height:
        root.height = 1 + max(self.getHeight(root.left), self.getHeight(root.right))
        balanceFactor = self.get_balance(root)

        #balance the tree:
        if balanceFactor > 1:
            if self.get_balance(root.left) >= 0:
                return self.rightRotate(root)
            else:
                root.left = self.leftRotate(root.left)
                return self.rightRotate(root)
        if balanceFactor < -1:
            if self.get_balance(root.right) <= 0:
                return self.leftRotate(root)
            else:
                root.right = self.rightRotate(root.right)
                return self.leftRotate(root)
        return root
    
    #function to perform left rotation:
    def leftRotate(self, z):
        y = z.right
        T2 = y.left
        y.left = z
        z.right = T2
        z.height = 1 + max(self.getHeight(z.left), self.getHeight(z.right))
        y.height = 1 + max(self.getHeight(y.left), self.getHeight(y.right))
        return y

    #right rotation:
    def rightRotate(self, z):
        y = z.left
        T3 = y.right
        y.right = z
        z.left = T3
        z.height = 1 + max(self.getHeight(z.left), self.getHeight(z.right))
        y.height = 1 + max(self.getHeight(y.left), self.getHeight(y.right))
        return y

    #get the height of the node:
    def getHeight(self, root):
        if not root:
            return 0
        return root.height
    #get balance factor of the node:
    def get_balance(self, root):
        if not root:
            return 0
        return (self.getHeight(root.left) - self.getHeight(root.right))
    #def getMinValNode:
    def getMinValNode(self, root):
        if root.left is None or root.right is None:
            return root
        return self.getMinValNode(root.left)

    def preOrder(self, root):
        if not root:
            return
        print("{0} ".format(root.key), end="")
        self.preOrder(root.left)
        self.preOrder(root.right)
    
    #print helper:
    def printHelper(self, currPtr, indent, last):
        if currPtr != None:
            sys.stdout.write(indent)
            if last:
                sys.stdout.write("R---")
                indent +="   "
            else:
                sys.stdout.write("L---")
                indent+="|   "
            print(currPtr.key)
            self.printHelper(currPtr.left, indent, False)
            self.printHelper(currPtr.right, indent, True)

if __name__ == "__main__":
    myTree = AVLTree()

    root = None

    nums = [33, 13, 52, 9, 21, 61, 8, 11]

    #populate the tree:
    for num in nums:
        root = myTree.insert_node(root, num)
    myTree.printHelper(root, "", True)

    key = 61
    root = myTree.delete_node(root, key)

    print("After deletion: ")

    myTree.printHelper(root, "", True)
    



        

File: Data Structures/Tree-based-DSAs/Binary-tree/avl-tree.cpp
/**
 * @file avl-tree.cpp
 * @author Lenny Orengo
 * @brief avl tree cpp implementation
 * @version 0.1
 * @date 2022-11-21
 */
#include "iostream"

using namespace std;

/* avl tree class: */
class Node{
    public:
        int key;
        Node * left;
        Node * right;
        int height;
};

/* max num math helper:*/
int max(int a, int b){
    return (a > b )? a : b;
}

/* get height: */

int height(Node * n){
    if (!n){
        return 0;
    }
    return n->height;
}

/*new node creation */
Node * newNode(int key){
    Node * node = new Node();
    node->key = key;
    node->left = nullptr;
    node->right = nullptr;
    node->height = 1;
    return node;
}

/* right rotate: */

Node * rightRotate(Node * y){
    Node * x = y->left;
    Node * T2 = x->right;
    x->right = y;
    y->left = T2;
    y->height = max(height(y->left), height(y->right)) + 1;
    x->height = max(height(x->left), height(x->right)) + 1;
    return x;
}

/* rotate left: */

Node * leftRotate(Node * x){
    Node * y = x->right;
    Node * T3 = y->left;
    y->left = x;
    x->right = T3;
    y->height = max(height(y->left), height(y->right)) + 1;
    x->height = max(height(x->left), height(x->right)) + 1;
    return y;
}

/* get balance factor of each node:*/

int getBalanceFactor(Node * n){
    if(!n){
        return 0;
    }
    return height(n->left) - height(n->right);

}

/* insert node */

Node * insertNode(Node * node, int key){
    /* find the correct pos and insert the node:*/

    if (!node){
        return newNode(key);
    }

    if (key < node->key){
        node->left = insertNode(node->left, key);
    }
    else if (key > node->key){
        node->right = insertNode(node->right, key);
    }
    return node;
    /* update the balance factore of each node and balance the tree:*/
    node->height = 1 + max(height(node->left), height(node->right));

    int balanceFactor = getBalanceFactor(node);
    if (balanceFactor > 1){
        if (key < node->left->key){
            return rightRotate(node);
        }
        else if (key > node->left->key){
            node->left = leftRotate(node);
            return rightRotate(node);
        }
    }
    if (balanceFactor < -1){
        if (key > node->right->key){
            return leftRotate(node);
        }
        else if (key < node->right->key){
            node->right = rightRotate(node->right);
            return leftRotate(node);
        }
    }
    return node;

}
/* node with minimum value:*/

Node * nodeWithMinVal(Node * node){
    Node * current = node;
    while (current->left){
        current = current->left;
    }
    return current;
}

/* deleting a node:*/

Node * deleteNode(Node * root, int key){
    /* find the node and delete it:*/

    if (!root){
        return root;
    }
    if (key < root->key){
        root->left = deleteNode(root->left, key);
    }
    else if (key > root->key){
        root->right = deleteNode(root->right, key);
    }
    else{
        if (!root->left || !root->right){
            Node * temp = root->left ? root->left : root->right;
            if (!temp){
                temp = root;
                root = nullptr;
            } else{
                *root = *temp;
                free(temp);
            }
        } else{
            Node * temp = nodeWithMinVal(root->right);
            root->key = temp->key;
            root->right = deleteNode(root->right, temp->key);
        }
    }

    if (!root){
        return root;
    }

    /* update the balance factor of each node and balance the tree:*/
    
}